File: 45-jump-game-ii/jump-game-ii.java
class Solution {

    int helper(int curr, int[] nums,int[] dp)
    {
        if(curr >= nums.length - 1)
        {
            return 0;
        }

        if(dp[curr] != -1)
        {
            return dp[curr];
        }

        int ans = Integer.MAX_VALUE;

        for(int i=1;i<=nums[curr];i++)
        {
            int next = Integer.MAX_VALUE;
            next = helper(curr + i,nums,dp);

            if(next != Integer.MAX_VALUE)
            {
                ans = Math.min(ans,1+next);
            }
        }
        dp[curr] = ans;
        return dp[curr];
    }

    public int jump(int[] nums) {

       int dp[] = new int [nums.length + 1];
        Arrays.fill(dp,-1);
        return helper(0,nums,dp);
    }
}

File: 45-jump-game-ii/jump-game-ii.cpp
class Solution {
public:
    void helper(int curr, vector<int>& nums, vector<int>& dp, int& ans) {
        if (curr >= nums.size() - 1) {
            ans = 0;
            return;
        }

        if (dp[curr] != -1) {
            ans = dp[curr];
            return;
        }

        int min_jumps = INT_MAX;

        for (int i = 1; i <= nums[curr]; i++) {
            int next_ans = INT_MAX;
            helper(curr + i, nums, dp, next_ans);

            if (next_ans != INT_MAX) {
                min_jumps = min(min_jumps, 1 + next_ans);
            }
        }

        dp[curr] = min_jumps;
        ans = dp[curr];
    }

    int jump(vector<int>& nums) {
        vector<int> dp(nums.size(), -1);
        int ans = INT_MAX;
        helper(0, nums, dp, ans);
        return ans;
    }
};
